feat(textarea): 718 - Add Textarea component

Jacksonmills opened the pull request:
Add the new `Textarea`(#718) component to provide an unstyled native text input with Field system integration:

* Exposes `<Textarea … />` with support for controlled and uncontrolled usage.
* Integrates with the **Field** system for automatic state handling and validation.
* Supports content-aware auto-resizing via `minRows` and `maxRows`.
* Exposes data attributes for styling based on control state.
* Sets up CSS Modules and Tailwind usage examples.
* Includes unit tests for core behavior.
* Ships a demo page under `docs/src/app/(public)/content/react/components/textarea`.

This is my first time adding a form control to base-ui. I’m following the repo’s patterns and totally down to adjust if this component needs tweaks.

**Changed files**

* `packages/react/src/textarea/Textarea.tsx`
* `packages/react/src/textarea/index.ts`
* `docs/src/app/(public)/content/react/components/textarea/*`
* `packages/react/src/textarea/Textarea.test.tsx`

---

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/base-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Jacksonmills committed:
feat(textarea): 789012 - Add Textarea component

Add a native, unstyled textarea component that integrates with the Field system to provide automatic state handling and validation.

Enable content-aware auto-resizing via minRows/maxRows, support controlled and uncontrolled usage, and expose data attributes for styling based on control state. Add documentation, API reference, unit tests, and package export so the component is discoverable and consumable.

Jacksonmills committed:
add demos

Jacksonmills committed:
Merge branch 'master' into feat/718-Textarea

Jacksonmills committed:
revert form demo change

Jacksonmills committed:
Merge branch 'feat/718-Textarea' of github.com:Jacksonmills/base-ui into feat/718-Textarea

Jacksonmills committed:
prettier

Jacksonmills committed:
dedupe

Jacksonmills committed:
stylelint

Jacksonmills committed:
Merge branch 'master' into feat/718-Textarea

colmtuite commented on packages/react/src/textarea/Textarea.tsx:
```diff
@@ -0,0 +1,228 @@
+
+    const resize = () => {
+      el.style.height = 'auto';
+      const lineHeight = parseFloat(getComputedStyle(el).lineHeight) || 24;
+      const min = minRows != null ? minRows * lineHeight : 0;
+      const maxPx = maxRows != null ? maxRows * lineHeight : Infinity;
+      const target = Math.min(maxPx, Math.max(min, el.scrollHeight));
```
Perhaps some research into when `field-sizing` might be at 90%+ support would be helpful too. When it's widely supported, we won't need this functionality in JS. https://caniuse.com/?search=field-sizing

Jacksonmills replied:
Absolutely, would be great to drop the JS once browser support crosses the 90% mark. Do you have a usual way to track when features hit that kind of support? Just Can I Use, or anything more automated you follow?

colmtuite replied:
nothing automated

Jacksonmills replied:
Alright I was able to align things more with MUI's logic

Also added a dedicated demo for field-sizing explaining the browser support

Jacksonmills committed:
Remove field-sizing from examples

Removes usage of the experimental field-sizing property and related Tailwind utility from textarea examples and removes hardcoded min-height in CSS modules.

Avoids depending on a feature with limited browser support (~71.47%). Adds space to create a dedicated example and documentation with compatibility research once support reaches ~90%, keeping demos stable and predictable in the meantime.

Jacksonmills committed:
update docs

Jacksonmills committed:
Merge branch 'feat/718-Textarea' of github.com:Jacksonmills/base-ui into feat/718-Textarea

Jacksonmills committed:
Remove 'compact' attribute from demo components in textarea documentation

Jacksonmills committed:
Enhance Textarea component with automatic height resizing and hidden textarea for improved performance

Jacksonmills committed:
Extracts height calc and debounces resize

Extracts the textarea height calculation into a reusable utility and isolates the expensive measurement logic.

Reduces layout thrashing by tracking last-applied height and overflow to avoid redundant DOM writes, and introduces debounced scheduling (~166ms) for resize/observer events while using immediate RAF for input events to keep input responsive.

Adds proper timer/observer cleanup and a default rows fallback to ensure consistent initial sizing and improved runtime performance.

Jacksonmills committed:
Improve textarea style copying

Replace a hardcoded property list and unsafe direct assignments with iteration over computed styles and setProperty/getPropertyValue to copy relevant CSS keys.

This avoids TypeScript readonly casts, picks up actual computed property names (including vendor-prefixed variants), and makes style copying more robust and cross-browser resilient.

Jacksonmills committed:
adopting the MUI Firefox double-set workaround, leading to more stable textarea height calculations.

Jacksonmills committed:
Remove forced resize none

Remove built-in inline "resize: none" from textarea rendering.

Allow consumers to control resizing via their own CSS instead of enforcing non-resizable behavior. Preserve existing value/rows handling; change affects styling only.

Jacksonmills committed:
moved resize none to example CSS

Jacksonmills committed:
Add field-sizing demo

Adds interactive demos showing CSS field-sizing for the textarea component using CSS Modules and Tailwind, and updates documentation to explain the feature, browser support, and recommended fallback guidance.

Provides examples to let developers try native field-sizing behavior, clarifies current limited browser support, and notes a plan to adopt field-sizing by default once support reaches sufficient coverage.

Jacksonmills committed:
dedupe

Jacksonmills committed:
cleanup comments

mnajdova commented:
@Jacksonmills this is a great first iteration. The team is a bit busy this week with React Conf, but we can come back to this in the next weeks. Thanks for putting in the effort :)

Jacksonmills commented:
> @Jacksonmills this is a great first iteration. The team is a bit busy this week with React Conf, but we can come back to this in the next weeks. Thanks for putting in the effort :)

ofc no worries!

Jacksonmills committed:
Merge branch 'master' into feat/718-Textarea

Jacksonmills committed:
refactor(textarea): update demo imports and fix form integration

Updates demo imports to use named exports for improved modularity and replaces deprecated form field props to align with recent API changes. Ensures demo rendering and form validation remain consistent with updated field logic.

michaelhazan commented:
hey!
are there any updates on this?
any way to help with it?

simka commented:
Hey! Have anything moved forward over the last 2 months?

mj12albert commented:
Sorry for not getting back to this sooner, we decided to not have this component
